### PR TITLE
[BEAM-6017] [WIP] Support nanosecond precision for DateTime field in Row

### PR DESCRIPTION
**Pull request by amaliujia**
Support nanosecond precision for DateTime field in Row.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by amaliujia** [comment]
@reuvenlax @kennknowles

**Comment by amaliujia** [comment]
run java postcommit

**Comment by amaliujia** [comment]
Run Dataflow ValidatesRunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 1]
```diff
@@ -283,8 +292,12 @@ public String getString(int idx) {
    */
   @Nullable
   public ReadableDateTime getDateTime(int idx) {
-    ReadableInstant instant = getValue(idx);
-    return instant == null ? null : new DateTime(instant).withZone(instant.getZone());
+    Instant instant = getValue(idx);
```
Comment (amaliujia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: I think `public abstract <T> T getValue(int fieldIdx);` should be changed to `private` and all value access to Row should go though specific type access method, e.g. `public int16 getInt6(int idx)`, `public ReadableDateTime getDateTime`, etc. 

By doing so, we can hide the internal representation of Row. 

Right now I at least see BeamSQL implementation uses `public <T> T getValue(int fieldIdx)` directly, which access the value saved in Row.

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslUdfUdafTest.java [hunk 2]
```diff
@@ -91,6 +84,23 @@ public void testJodaTimeUdfUdaf() throws Exception {
     pipeline.run().waitUntilFinish();
   }
 
+  @Ignore("Fix the access to DateTime field of Row in Udaf")
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslUdfUdafTest.java]: Current change breaks user defined aggregation function. Need to fix it before merging PR.

**Comment by reuvenlax** [comment]
I'll look at this PR. It might require updating the ByteBuddy code.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java [hunk 3]
```diff
@@ -185,6 +186,12 @@
               keyType,
               valueType,
               setterFactory);
+    } else if (type.getTypeName().equals(TypeName.DATETIME)) {
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java]: I don't think this is the correct place to put this. If anything the conversion should happen inside Row instead.

**Comment by mouyang** [comment]
It would be nice to restart discussions about adding this functionality in Beam 2.x because BigQueryUtils currently drops the microseconds?  In order to maintain backwards compatibility and facilitate DATETIME removal in 3.x, it might be better to create a new type altogether in the interim.

**Pull request closed without merging**